File: apps/dice.py
"""
TinyOLED Desktop — Zar Simülatörü
UP/DOWN: Zar tipi, SEL: At, LONG: Çıkış
"""
import random, time
from core.framebuffer import Framebuffer

DICE_TYPES = [4, 6, 8, 10, 12, 20]

class DiceApp:
    NAME = "dice"; LABEL = "Zar"; ICON = "dice"

    def __init__(self, on_exit):
        self.on_exit = on_exit; self._type_idx = 1; self.result = 0
        self._rolling = False; self._roll_frames = 0; self._last = 0

    def on_up(self): self._type_idx = (self._type_idx - 1) % len(DICE_TYPES)
    def on_down(self): self._type_idx = (self._type_idx + 1) % len(DICE_TYPES)
    def on_sel(self):
        if not self._rolling:
            self._rolling = True; self._roll_frames = 0
    def on_long(self): self.on_exit()

    def update(self):
        if not self._rolling: return
        now = time.monotonic()
        if now - self._last < 0.05: return
        self._last = now
        self._roll_frames += 1
        sides = DICE_TYPES[self._type_idx]
        self.result = random.randint(1, sides)
        if self._roll_frames >= 20: self._rolling = False

    def draw(self, fb: Framebuffer):
        sides = DICE_TYPES[self._type_idx]
        fb.text(f"Zar: D{sides}", 2, 10)
        fb.hline(0, 18, 128)
        # Zar tipi göstergesi
        for i, d in enumerate(DICE_TYPES):
            x = 4 + i * 21; sel = (i == self._type_idx)
            if sel: fb.rect(x-1, 20, 20, 10, fill=True)
            fb.text(f"D{d}", x, 21, on=not sel)
        # Büyük sonuç
        if self.result > 0:
            r_str = str(self.result)
            scale = 4
            total_w = len(r_str) * 6 * scale
            sx = (128 - total_w) // 2
            for ch in r_str:
                from core.font import Font
                cols = Font.glyph(ch)
                for ci, col in enumerate(cols):
                    for ri in range(7):
                        if col & (1 << ri):
                            fb.rect(sx+ci*scale, 34+ri*scale, scale, scale, fill=True)
                sx += 6 * scale
        if self._rolling:
            fb.text_centered("Yuvarlanıyor...", 56)
        else:
            fb.text("[SEL] At", 2, 56)

File: web-simulator/js/apps/dice.js
/** TinyOLED Desktop — Dice Simulator */
import { Font } from '../font.js';
const DICE_TYPES=[4,6,8,10,12,20];

export class DiceApp {
  static NAME='dice'; static LABEL='Zar'; static ICON='dice';
  constructor(onExit) {
    this.onExit=onExit; this._typeIdx=1; this.result=0;
    this._rolling=false; this._rollFrames=0; this._last=0;
  }
  onUp()   { this._typeIdx=(this._typeIdx-1+DICE_TYPES.length)%DICE_TYPES.length; }
  onDown() { this._typeIdx=(this._typeIdx+1)%DICE_TYPES.length; }
  onSel()  { if(!this._rolling){this._rolling=true;this._rollFrames=0;} }
  onLong() { this.onExit(); }

  update() {
    if(!this._rolling) return;
    const now=performance.now()/1000;
    if(now-this._last<0.05) return;
    this._last=now; this._rollFrames++;
    this.result=1+Math.floor(Math.random()*DICE_TYPES[this._typeIdx]);
    if(this._rollFrames>=20) this._rolling=false;
  }

  draw(fb) {
    const sides=DICE_TYPES[this._typeIdx];
    fb.text(`Zar: D${sides}`,2,10); fb.hline(0,18,128);
    for(let i=0;i<DICE_TYPES.length;i++) {
      const x=4+i*21, sel=(i===this._typeIdx);
      if(sel) fb.rect(x-1,20,20,10,true,true);
      fb.text(`D${DICE_TYPES[i]}`,x,21,!sel);
    }
    if(this.result>0) {
      const rStr=String(this.result), scale=4;
      let sx=Math.floor((128-rStr.length*6*scale)/2);
      for(const ch of rStr) {
        const cols=Font.glyph(ch);
        for(let ci=0;ci<cols.length;ci++)
          for(let ri=0;ri<7;ri++)
            if(cols[ci]&(1<<ri)) fb.rect(sx+ci*scale,34+ri*scale,scale,scale,true,true);
        sx+=6*scale;
      }
    }
    fb.text(this._rolling?'Yuvarlaniyor...':'[SEL] At',2,56);
  }
}
